### ISSUE TITLE
[a11y Support] CAIA - CodeSpace Review with Assistive Technology

### ISSUE THREAD

**Issue by sara-amanda** [department-of-veterans-affairs/va.gov-team#94060]
> [!NOTE]
> **QUICK SUMMARY:** CodeSpace was mentioned as a testing environment during the Research, Design and Content weekly meeting on 9/30/2024. Platform recommends CodePen or Staging at this time. CAIA wants to test CodeSpace for Accessibility. 
> - **Past use case:** CodeSpace was already running, click accept to bypass a login for Medical Records. No known issues encountered, per Riley. 
> - **Recent Backend AT Testing (Developer Side):**  The hardest part would be bypassing that and other barriers. Until you get to that point you have to use a lot of non-traditional keystrokes. 
> - **New AT Prototype Testing (User Side):** `Completed` (see ticket comments)

### Testing Environments & Assistive Technology
During the Research, Design and Content Sync on 9/30/2024, the team presenting, the Authenticated Experience: Design Patterns Team, mentioned that they used CodeSpace for testing the prototype, instead of CodePen.
- There was an evaluation comparing the two testing environments, per Adam Whitlock in the OCTO #design Slack discussion:
   - _"Here is the authed patterns [discovery work around codepen vs codespaces](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/authenticated-patterns/engineering/%2332-Discover%20and%20Evaluate%20Options%20for%20Hosting%20Experimental%20'Pre-fill'%20Pattern%20Work.md) (it also goes through several other options that we looked at as well)."_
- The team has not yet used CodeSpace with assistive technology users, that will be the next round of testing. So they have not experienced any issues yet.

## Initial Review & Feedback

**CAIA would like to test a prototype with CodeSpace, so we took a gander, and here is our initial* feedback:** 

- **Conclusion 1:** CodePen or Staging are the preferred environments to test in, when it comes to AT users. On the technical building side of things, CodeSpace appears to introduce barriers on the developer side. 
- **Conclusion 2:** `See ticket comments for results`
    - **Next Steps:** CodeSpace has been used in the Medical Records Study for MHV and worked. We will test with the CodeSpace version for Auth. Experience: Design Patterns, to check for any issues or documentation needs when using CodeSpace.
- **Rationale 1:** While CodeSpace detects a screen reader, it is clunky when it switches between forms mode and browse mode, when reviewing what is on the screen. Meaning, that most typical screen reader keystrokes will not work. Keystrokes are what a screen reader user will be heavily reliant upon when navigating. 
- **Rational 2:** `See ticket comments for results`

## About the Team

- **Team name:** CAIA 
- **Product Testing:** CodeSpace

### Supporting Artifacts 
- Auth. Exp.: Design Patterns Team's [discovery work around codepen vs codespaces](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/authenticated-patterns/engineering/%2332-Discover%20and%20Evaluate%20Options%20for%20Hosting%20Experimental%20'Pre-fill'%20Pattern%20Work.md) (it also goes through several other options that we looked at as well)
- [Prototype to Test] (Link Needed)
- [Monday Design Meeting 9/30/2024 - Teams Recording](https://dsva.slack.com/archives/C0NGDDXME/p1727736591911269?thread_ts=1727701208.628089&cid=C0NGDDXME)
- [Monday Design Meeting 9/30/2024 - Slides PDF](https://dsva.slack.com/archives/C0NGDDXME/p1727720699381979?thread_ts=1727701208.628089&cid=C0NGDDXME)

```[tasklist]
### a11y Tasks
- [x] **Collaborate and discuss** CodeSpace testing with users of Assistive Tech
- [x] **Request prototype link** from Auth Exp Design Pattern Team [10/1/2024](https://dsva.slack.com/archives/C07PXQEQU3U/p1727812226229109)
- [x] **Receive prototype link** from Auth Exp Design Pattern Team `Date TBD`
- [x] **a11y Support:** Test CodeSpace environment with Assistive Tech @mutiaadhocteam
- [x] **Document** how to best use CodeSpace for AT testing and/or consideration for users of AT when using CodeSpace
- [x] **Review** feedback and documentation and share with Martha
- [x] **Create ticket with Platform Content Team** `10/8/2024`  #94575
```
```[tasklist]
### Acceptance Criteria
- [x] CodeSpace is tested with AT (externally - as a developer) @mutiaadhocteam
- [x] CodeSpace is tested with AT (internally - as a user) @mutiaadhocteam
- [x] CodeSpace/AT documentation determination need is made (yes/no)
- [x] Martha has reviewed this work
- [x] **GitHub:** Update the [menu of services for accessibility](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/teams/CAIA/accessibility#staging--codepen--production-audit) to include CodeSpace @coformajamie
- [ ] **Platform Site** - Updates Completed: ref. ticket #94575 - @sara-amanda (Req. Opened Tix 10/8/2024) in Backlog ready for Platform Content Team's Review as of 10/9/2024
```
```[tasklist]
## Ticket Updates (CAIA Internal)
- [x] _Connect to an `Epic` or `Intake` (what body of work is this a part of?)_
- [x] _Label with `Originator/Team` (product team or stakeholder requesting support)_
- [x] _Label date in the `Open Date` field_
- [x] _Label with `Estimate` (level of effort expected for this ticket)_
- [x] _Add `Assignee(s)` name(s) to ticket_ (include Naomi)
- [x] _Add `Assignee(s) name(s) to each task` they will complete via handle tag (if known)_
- [x] _Select a `Priority Level`_
- [x] _Update date in `Last Checked` field_
- [ ] _Label with `Actual` (level of effort it took to complete this ticket)_
- [ ] _Update date in_ `Closed Date`
- [ ] Move from `Validate` to `Closed` Status (once acceptance criteria is met)
```

```[tasklist]
### Research Studies with CodeSpace Prototypes
- [ ] #90140
- [ ] #90780
- [ ] https://github.com/department-of-veterans-affairs/va.gov-team/issues/94575
```

**Comment by sara-amanda** [comment]
Added two study examples that used CodeSpace to this ticket for quick ref.

**Comment by mutiaadhocteam** [comment]
**TASK:** _Document how to best use CodeSpace for AT testing and/or consideration for users of AT when using CodeSpace._

_How do we best use CodeSpace for AT testing?_

- The link to the environment should be provided to the AT user beforehand.
- Ideally, any login flow or extraneous steps to access the environment would be bypassed to allow the user to test the environment with as little barriers as possible.
- Currently, no issues in regard to AT testing with CodeSpace have been experienced or reported. The software behavior of the AT within the CodeSpace environment on the testing side reflects that of the typical behavior of the AT outside of the testing environment.

_What are other considerations for users of AT when using CodeSpace for testing?_

- Ensure that the user is aware of how to exit CodeSpace after the testing is complete.
- As the software behavior of the AT within the CodeSpace environment reflects that of the typical software behavior of the AT outside of the testing environment, it is important to communicate to the user when / if they have exited the CodeSpace environment.

**Comment by mutiaadhocteam** [comment]
**Extra Information:** _What are the concerns of using CodeSpace with AT on the developer side?_

While CodeSpace detects a screen reader, it is clunky when it switches between forms mode and browse mode, when reviewing what is on the screen. Meaning, that most typical screen reader keystrokes will not work. Keystrokes are what a screen reader user will be heavily reliant upon when navigating.

_Finding: Typical screen reader keystrokes may not work until full-screen mode is enabled._

- Blind and visually impaired users may have difficulty understanding if full-screen mode is enabled. Typical screen reader keystrokes may not work until full-screen mode is enabled.
- The software behavior of the AT when using CodeSpace on the developer side may not mimic the typical software behavior of the AT outside of CodeSpace. As many users rely on screen reader keystrokes for navigation and comprehension of the screen's contents and structure
- However, there are many known limitations of using CodeSpace with keybindings and workarounds for the desktop.
- A few solutions are listed [here](https://code.visualstudio.com/docs/remote/codespaces).

**While CodeSpace can be downloaded as a Progressive Web App(PWA) to help provide native-like experiences to users on supporting devices, the need to customize keystrokes specifically with CodeSpace and a screen reader running at the same time still exists. Additionally, as noted in this [GitHub forum](https://github.com/orgs/community/discussions/54040), there still exist multiple gaps in functionality when using a screen reader with CodeSpace.**

**Comment by sara-amanda** [comment]
# Next Steps: Ready for Martha

@artsymartha68 the findings above are ready for your review c/o @mutiaadhocteam who reviewed CodeSpace from both the user testing side and the developer side.  Our questions for you and possible next steps, can be found below.

## Menu of Services

- **Would you like us to update the [menu of services for accessibility](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/teams/CAIA/accessibility#staging--codepen--production-audit) to include CodeSpace?** 
    - _The menu is [linked from the VA Platform page for CAIA under the Collaboration Cycle](https://depo-platform-documentation.scrollhelp.site/collaboration-cycle/sitewide-content-and-ia-intake-request#Content,Accessibility,andInformationArchitecturesupport-SitewideCAIA-Accessibility) to the GitHub page._ 
        - [ ] Yes, update the menu of services for accessibility
        - [ ] No, do not update the menu of services for accessibility

## Research Documentation
- **There is also reference to CodePen in the Research at VA section of the Platform site.** 
    - **Path**: [Research at VA / Remote Moderated Research / Assistive Technology / Accessible Prototyping with CodePen](https://depo-platform-documentation.scrollhelp.site/research-design/accessible-prototyping-with-codepen)
    - Would you like to recommend an additional note about Codespace here, based on the feedback above from Maria? 
        - [ ] Yes, share feedback with the research team. 
        - [ ] No, do not share feedback with the research team. 

## Developer Docs
- [GitHub Codespaces are referenced under Developer Docs](https://depo-platform-documentation.scrollhelp.site/developer-docs/using-github-codespaces) on the VA Platform site.
    - Would you like for us to recommend an additional note about Codespace here, based on the feedback above from Maria? 
        - [ ] Yes, share feedback with Platform for the Developer Docs. 
        - [ ] No, do not share feedback with Platform for the Developer Docs.

**Comment by artsymartha68** [comment]
Thank you so much for this thorough analysis and making it so easy for me to review! Let's move forward with updating all the documentation locations, that makes sense to me since there is a groundswell already building for the CodeSpace option.

**Comment by coforma-jamie** [comment]
CAIA menu of services has been updated ✅

**Comment by sara-amanda** [comment]
## Platform Site Updates
_Updates as of 10/9/2024_

**Platform received our request for updates:** 
- **Current Status:** `Backlog` 🟡 
- **Next Steps:** Review by the Platform Content Team and next steps
- **Platform Ticket:** https://github.com/department-of-veterans-affairs/va.gov-team/issues/94575#issuecomment-2402935205

**Comment by sara-amanda** [comment]
## CAIA Ticketing Process
@NaomiPMC - since this is a unique ticket, with a support ticket tied to it ... 
- Would you like us to keep this in `validate` until the Platform Content Team makes their updates
- Or, would you prefer I `close` this out and we monitor the Platform ticket, for the remaining work? 

Please advise.

**Comment by sara-amanda** [comment]
> ## Platform Site Updates
> 
> _Updates as of 10/9/2024_
> 
> **Platform received our request for updates:**
> 
> * **Current Status:** `Backlog` 🟡
> * **Next Steps:** Review by the Platform Content Team and next steps
> * **Platform Ticket:** [Content Updates to Platform Site - Regarding Codespace (Post-Audit c/o CAIA Accessibility) #94575 (comment)](https://github.com/department-of-veterans-affairs/va.gov-team/issues/94575#issuecomment-2402935205)

- [x] @NaomiPMC closing this ticket, as the work is being tracked in the Platform Content Team's Backlog

**Comment by adamwhitlock1** [comment]
@coforma-jamie Here are the [docs for editing code on a codespace, but directly from a developer's IDE](https://docs.github.com/en/codespaces/developing-in-a-codespace/using-github-codespaces-in-visual-studio-code) as if it was on local drive.

I think this would be a good thing to add as it seems to be the most accessible way to access a codespace for a developer
